txnkv/transaction/pessimistic.go

```diff
@@ -150,156 +168,375 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
+				return err
+			}
+			if finished {
+				return nil
 			}
-			same, err := batch.relocate(bo, c.store.GetRegionCache())
+		} else if action.wakeUpMode == kvrpcpb.PessimisticLockWakeUpMode_WakeUpModeForceLock {
+			finished, retryMutations, err := action.handlePessimisticLockResponseForceLockMode(c, bo, &batch, mutations, resp, &diagCtx)
 			if err != nil {
 				return err
 			}
-			if same {
-				continue
+			if finished {
+				return nil
+			}
+
+			if retryMutations != nil && retryMutations.Len() < batch.mutations.Len() {
+				// The request may be partially succeeded. Replace the request with remaining mutations.
+				// The primary lock (if it's in this request) must have been succeeded.
+				batch = batchMutations{
+					region:    batch.region,
+					mutations: retryMutations,
+					isPrimary: false,
+				}
+				m = batch.mutations
+				mutations = convertMutationsToPb(m)
+				req.PessimisticLock().Mutations = mutations
 			}
-			err = c.pessimisticLockMutations(bo, action.LockCtx, batch.mutations)
-			return err
 		}
-		if resp.Resp == nil {
-			return errors.WithStack(tikverr.ErrBodyMissing)
+
+		// Handle the killed flag when waiting for the pessimistic lock.
+		// When a txn runs into LockKeys() and backoff here, it has no chance to call
+		// executor.Next() and check the killed flag.
+		if action.Killed != nil {
+			// Do not reset the killed flag here!
+			// actionPessimisticLock runs on each region parallelly, we have to consider that
+			// the error may be dropped.
+			if atomic.LoadUint32(action.Killed) == 1 {
+				return errors.WithStack(tikverr.ErrQueryInterrupted)
+			}
 		}
-		lockResp := resp.Resp.(*kvrpcpb.PessimisticLockResponse)
-		keyErrs := lockResp.GetErrors()
-		if len(keyErrs) == 0 {
-			if action.LockCtx.Stats != nil {
-				action.LockCtx.Stats.MergeReqDetails(reqDuration, batch.region.GetID(), sender.GetStoreAddr(), lockResp.ExecDetailsV2)
+	}
+}
+
+func (action actionPessimisticLock) handlePessimisticLockResponseNormalMode(c *twoPhaseCommitter, bo *retry.Backoffer, batch *batchMutations, mutationsPb []*kvrpcpb.Mutation, resp *tikvrpc.Response, diagCtx *diagnosticContext) (finished bool, err error) {
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return true, err
+	}
+	if regionErr != nil {
+		// For other region error and the fake region error, backoff because
+		// there's something wrong.
+		// For the real EpochNotMatch error, don't backoff.
+		if regionErr.GetEpochNotMatch() == nil || locate.IsFakeRegionError(regionErr) {
+			err = bo.Backoff(retry.BoRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				return true, err
+			}
+		}
+		same, err := batch.relocate(bo, c.store.GetRegionCache())
+		if err != nil {
+			return true, err
+		}
+		if same {
+			return false, nil
+		}
+		err = c.pessimisticLockMutations(bo, action.LockCtx, action.wakeUpMode, batch.mutations)
+		return true, err
+	}
+	if resp.Resp == nil {
+		return true, errors.WithStack(tikverr.ErrBodyMissing)
+	}
```

Comment:
There seem to be some duplicates with the implementation in `handlePessimisticLockResponseForceMode`, could it be abstracted to a single function?

Reply:
Ok. Btw I just realize that some of the code to handle the partially-successful cases is not removed. I'll remove them.